[MRG+1] Fix lower_bound_ not equal to max lower bound in mixture models when n_init > 1

ageron opened the pull request:
#### Reference Issues/PRs
Fixes #10869

#### What does this implement/fix? Explain your changes.
Just set the `lower_bound_` to be equal to the `max_lower_bound` at the end of the loop over the initializations (at the end of `BaseMixture.fit()`).
Also fix the `test_init()` function that was supposed to catch this bug. I do this by looping over multiple random states rather than just trying one (which had a 50% chance of wrongly succeeding).

#### Any other comments?
Thanks for making such a great library! :)

ageron committed:
Set lower_bound_ to max lower bound at the end of BaseMixture.fit(), fixes #10869

jnothman reviewed (commented):
I think this is right, but it makes the repeated setting of lower_bound_ look inexplicable (unless it were intended for debugging a crash!). Can we use a local lower_bound instead?

ageron commented:
Hi @jnothman ,
Good point.  I just made `lower_bound` a local variable instead.

ageron commented:
Mmh, my fix revealed another issue in `test_monotonic_likelihood()`: it is now failing because of a tiny numerical precision error:

```
-2.6242953837251695 not greater than or equal to -2.624295383725169
```

The difference is -4.44e-16, I say we can ignore it. For example, I could modify this function to assert this:

```python
tol = 1e-10
assert_greater_equal(current_log_likelihood + tol, prev_log_likelihood)
```

Wdyt?

ageron committed:
Use a local lower_bound variable rather than self.lower_bound_ during training, in BaseMixture.fit()

ageron commented:
There were a couple traps, but I finally managed to get this thing to work. Here's the new behavior:

* When `warm_start` is `False` or it is the first time we fit the model, then:
    * the final `self.lower_bound_` is now the maximum lower bound across the multiple inits (fixes #10869),
    * and `self.converged_` is equal to `True` if *any* of the inits converged (unchanged behavior, but I wanted to highlight this fact, as it is somewhat surprising).
* When `warm_start` is `True` and it is not the first call to the `fit()` method, then:
    * `self.lower_bound_` is the max lower bound found during this call to the `fit()` method (even if it is lower than the max lower bound found earlier).
        * This is useful if you train a model on a first dataset, then you train it on another dataset: the lower bound may be lower after the second call to `fit()`, and that's expected.
    * Moreover, `self.converged_` is `True` if and only if the lower bound is within `tol` of the previous lower bound at any iteration. Upon the first iteration, `prev_lower_bound` is initialized to `self.lower_bound_` (the max lower bound from the previous run).
        * I had to implement this logic to allow the `test_monotonic_likelihood()` to work, since it uses `warm_start=True` and `max_iter=1`, and therefore the only way it can detect that it has converged is if it looks at `self.lower_bound_` from the previous iteration. There's a tiny risk that the algorithm has in fact not converged, but just happened to stumble upon a very close lower bound estimate. Maybe `prev_lower_bound` should be set to `self.lower_bound_` only when `max_iter=1`, or `-np.infty` otherwise?
Wdyt?

jnothman commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -970,14 +970,14 @@ def test_sample():
-    rand_data = RandomData(np.random.RandomState(random_state), scale=1)
-    n_components = rand_data.n_components
-    X = rand_data.X['full']
+    for random_state in range(100):
```
Could you double check that this runs quite quickly

ageron replied:
It runs in 4 seconds on my laptop. Is that sufficiently quick?  If not, we could reduce this to less than a second by iterating just 25 times. The probability for the unfixed code to pass this test would be 1/2^25, which is roughly 3e-8. Tell me what you prefer.

jnothman replied:
Yes, 4s for this kind of test is excessively long, IMO. And if the unfixed code only has a chance of 3e-8 of passing (did you get that the right way around?), then we can run it for fewer than that. Alternatively, we can reduce the size of X if that's not actually related to the property being tested.

ageron replied:
Okay, reducing to 25 iterations. I think I got the calculation right: the unfixed code sets the `lower_bound_` to the lower bound of the last initialization, which has 50% chance of being higher than the lower bound of the first initialization. So in order to pass the updated test, the unfixed code would need to have 50% chance 25 times in a row, which 0.5^25 = 3e-8. In other words, the new test would catch the bug in the unfixed code.

jnothman replied:
Sure, thanks. A smaller dataset would also be fine.

jnothman commented on sklearn/mixture/base.py:
```diff
@@ -191,6 +191,7 @@ def fit(self, X, y=None):
         X = _check_X(X, self.n_components, ensure_min_samples=2)
         self._check_initial_parameters(X)
 
+
```
Rm blank line

ageron replied:
Oops, indeed. Fixing this now.

jnothman reviewed (approved):
I think this logic looks reasonable. Do attribute docstrings need improvement?

ageron committed:
Remove extra empty line

ageron committed:
Update documentation and reduce test_init() iterations from 100 to 25

ageron commented:
I just updated the documentation, I hope it's clear.

I also changed the logic slightly: the previous `lower_bound_` is now only used to check for convergence when `warm_start` is `True` **and** `max_iter` is 1. Indeed, if `max_iter>1`, it's safer to check for convergence after the 1st iteration, in case the dataset has changed.

ageron commented:
Hi @jnothman , is there anything else you need from me to fix this bug or are we okay to merge?

jnothman reviewed (commented):
Can that new logic be tested?

This requires a second review before merge

jnothman commented:
Please add an entry to the change log at `doc/whats_new/v0.20.rst`. Like the other entries there, please reference this pull request with `:issue:` and credit yourself (and other contributors if applicable) with `:user:`

ageron committed:
Update whats_new/v0.20.rst to add mention of issue 10869, and reformat file to fit on 80 columns

ageron committed:
Remove extra line in whats_new/v0.20.rst

ageron commented:
Thanks for your feedback, @jnothman .  I just updated `doc/whats_new/v0.20.rst`, as requested.

jnothman commented:
I also asked: Can that new logic be tested?

ageron committed:
Add tests for convergence detection in Gaussian mixtures when warm_start=True

ageron commented:
Hi @jnothman , I just added the tests for the new logic. In short:

1. I added a test that convergence is properly detected when `warm_start=True`, with different values of `max_iter` (1, 2, and 50).
2. I added a test that convergence is never detected at the 1st iteration when `warm_start=True` and `max_iter > 1`. This ensures that no convergence will be detected by mistake at the first iteration if the dataset is changed.

ageron committed:
Remove unnecessary catch_exception blocks

ageron committed:
Merge branch 'master' into gaussian_mixture_lower_bound_fix

ageron committed:
Fix tests since n_iter_ was recently fixed to be increased by 1

ageron commented:
Hi @jnothman ,
I've updated this PR because there were some conflicting changes in `sklearn/mixture/base.py` regarding `n_iter_`.
Could you please confirm that everything looks good to you and merge this PR if it does?
Thanks!

jnothman commented on doc/whats_new/v0.20.rst:
```diff
@@ -26,6 +26,8 @@ random sampling procedures.
 - :class:`linear_model.OrthogonalMatchingPursuit` (bug fix)
 - :class:`metrics.roc_auc_score` (bug fix)
 - :class:`metrics.roc_curve` (bug fix)
+- :class:`mixture.BayesianGaussianMixture` (bug bix)
```
Remind me: does this pr actually change  the prediction? If it's only affecting the attribute value, I think we should leave it out of here

ageron replied:
Yes, it may change the prediction in some very rare case. Consider:

```python
gm = GaussianMixture(warm_start=True, max_iter=10)
gm.fit(X1)
gm.fit(X2)
```

If `X2` is different from `X1`, is does not make sense to use the previous value of `lower_bound_` as a starting point. The previous implementation would do that, and therefore it might wrongly detect convergence after the first iteration (this is unlikely, but possible, in particular if `X2` is very similar to `X1`, or if `tol` is large). This PR fixes this. Thus, the fixed algorithm might converge to a better solution and produce different (better) predictions.

You could consider these as separate issues: (1) wrong `lower_bound_` value, and (2) wrong convergence detection logic. However, they're both caused by the same few lines of code, so I fixed them both in this one PR.

jnothman replied:
Thanks for the explanation

jnothman commented on doc/whats_new/v0.20.rst:
```diff
@@ -212,8 +214,8 @@ Model evaluation and meta-estimators
 - Add `return_estimator` parameter in :func:`model_selection.cross_validate` to
-  return estimators fitted on each split. :issue:`9686` by :user:`Aurélien Bellet
-  <bellet>`.
+  return estimators fitted on each split. :issue:`9686` by
```
Argh! Please revert all changes unrelated to the present fix!

Please do not change unrelated things. It makes your contribution harder to review and may introduce merge conflicts to other pull requests.

ageron replied:
Okay, sorry about that, makes sense (FYI, I was trying to make the file use 80 characters per line).

jnothman commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -785,6 +784,51 @@ def test_warm_start():
+            gmm.fit(X)
+            if gmm.converged_:
+                break
+        assert_true(gmm.converged_)
```
Minor: since moving to pytest, we're trying to avoid such assert_* functions. Use a bare assert instead.

ageron replied:
Okay, fixing this now.

jnothman reviewed (commented):
This needs a review by someone else still.

ageron committed:
Revert changes unrelated to PR 10870 in doc/whats_new/v0.20.rst

ageron committed:
Replace assert_* with plain asserts because of the move to pytest

ageron committed:
Remove comment in whats_new/v0.20.rst that will be added upon merging

amueller commented on sklearn/mixture/base.py:
```diff
@@ -172,11 +172,18 @@ def _initialize(self, X, resp):
         trial, the method iterates between E-step and M-step for `max_iter`
         times until the change of likelihood or lower bound is less than
         `tol`, otherwise, a `ConvergenceWarning` is raised.
+        If `warm_start` is `True`, then `n_init` is ignored and a single
```
should be double backticks everywhere....

ageron replied:
Oh right, just fixed this, thanks.

amueller commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -785,6 +784,51 @@ def test_warm_start():
+    # lower_bound_ is very close to the lower_bound_ after the previous call
+    # to the fit method.
+    # Unlikely, but possible and problematic, so we might as well avoid it.
+    rng = np.random.RandomState(0)
```
Sorry I'm being slow. I don't see where this case is handled and I don't understand the test.  Why do we always reset with max_iter > 1 now?

ageron replied:
No worries, let me explain. The scenario I'm trying to avoid, which could happen today (without this PR), is using `warm_start=True`, and you run `gm.fit(X1)` followed by `gm.fit(X2)`, where `X1` and `X2` are different datasets, and if you are unlucky the first iteration of `gm.fit(X2)` happens to compute a lower bound very close to the final lower bound of `gm.fit(X1)`, so the algorithm thinks the algorithm has converged when it fact it should have continued to iterate.  Sure, this is quite unlikely with the default `tol` and if `X1` and `X2` are very different, but it might be dangerously likely if `X1` and `X2` are very similar, but not identical (or if `tol` is high).
Since it is hard to find two datasets `X1` and `X2` where this scenario occurs, I test this scenario by setting `tol` to infinity.
This case is handled on line 217 of `base.py`: we start the iterations with `lower_bound = (-np.infty if do_init or self.max_iter > 1 else self.lower_bound_)`. So we only continue from the final lower bound of the last call to `fit()` if `warm_start` is `True` and `max_iter == 1`. The assumption is that people who use `warm_start=True` and `max_iter = 1` are certainly doing this to manually run the training loop themselves on a single dataset, but if they are using `max_iter > 1`, it is unclear whether they are running consecutive calls to `fit()` on the same dataset or not, so we should err on the safe side.
So in short there are two issues in the current implementation: (1) if `n_init > 1`, the lower bound is the one from the last initialization, not from the best initialization, and (2) there is a risk of false convergence if `warm_start` is `True`.  Since they are both due to the same few lines of code, I fixed them both in this one PR.
Hope this helps.

ageron committed:
Replace single backticks with double backticks in doc string

ageron committed:
Limit is 79 characters per line, not 80.

jnothman commented:
I've not got a deep understanding yet, but I'm not sure we should be
focusing on users changing datasets between warm_start fits (which I
consider an abnormal use) or making implicit assumptions about the meaning
of max_iter=1

ageron commented:
@jnothman, thanks for your feedback.  I did not know that switching datasets between `warm_start` fits was an abnormal use, in fact I have often done this after updating a dataset with new data, to avoid starting training from scratch.  How else can this be done?  In any case, I feel that this will require more discussion, so if you like I can split this PR in two: one part will just focus on fixing the original bug (incorrect `lower_bound_` when `n_init > 1`) and another about the risk when switching datasets. Sounds good?

jnothman commented:
I see what you mean now. that's a good use. let's look again...

jnothman reviewed (commented):
I think the only concerning part here is that max_iter controls multiple things now. If it's not hard to split that from this PR, it would help us ensure that at least the bug is fixed for release

ageron commented:
Sure @jnothman , it shouldn't be too hard, I'll split the PR (probably this week-end).

ageron committed:
Merge branch 'master' into gaussian_mixture_lower_bound_fix

ageron committed:
Remove the false convergence fix to treat it in a separate PR

ageron commented:
Hi @jnothman , I updated this PR to keep only the fix for the original bug, i.e., the `lower_bound_` was the max `lower_bound_` of the last initialization (when `n_init > 1`) rather than the max `lower_bound_` across all initializations.
I'll file a separate bug and a separate PR for the wrong convergence detection when consecutive fits use different datasets and using warm start.

jnothman reviewed (approved):
Thanks. LGTM.

ageron commented:
Hi there, this PR needs a second review, whenever you have the chance.

GaelVaroquaux committed:
Merge branch 'master' into gaussian_mixture_lower_bound_fix

GaelVaroquaux commented:
I resolved the conflicts. I will merge once the tests pass.

GaelVaroquaux commented:
All tests pass aside from appveyor, which is lagging behind as an effect of the sprint.

Merging!

GaelVaroquaux commented:
Thank you!!!
